### DISCUSSION TITLE
using csrfGuard without SessionTokenKeyExtractor

### DISCUSSION THREAD

**Discussion by ACM-Unit** [OWASP/www-project-csrfguard#44] [Q&A] [answered]
HI,
could you please help me with next question?
Can I use csrfGuard with  4.1.1 version without using org.owasp.csrfguard.LogicalSessionExtractor = org.owasp.csrfguard.session.SessionTokenKeyExtractor
I see following comment in properties file : "Defaults to 'org.owasp.csrfguard.session.ContainerSession', which uses the container's HttpSession in the background"
but if I change it to org.owasp.csrfguard.LogicalSessionExtractor = org.owasp.csrfguard.session.ContainerSession I'll get exception

**Comment by forgedhallpass** [comment, marked as answer]
Hello @ACM-Unit,

Could you please let me know what exactly you are trying to achieve?

The [`LogicalSessionExtractor`](https://github.com/OWASP/www-project-csrfguard/blob/master/csrfguard/src/main/java/org/owasp/csrfguard/token/storage/LogicalSessionExtractor.java), as the name suggests is intended to **extract logical sessions** from requests. It has two methods that, return a [`LogicalSession`](https://github.com/OWASP/www-project-csrfguard/blob/master/csrfguard/src/main/java/org/owasp/csrfguard/session/LogicalSession.java) interface. The default implementation for the `LogicalSessionExtractor` is the [`SessionTokenKeyExtractor`](https://github.com/OWASP/www-project-csrfguard/blob/master/csrfguard-extensions/csrfguard-extension-session/src/main/java/org/owasp/csrfguard/session/SessionTokenKeyExtractor.java), which uses a [`ContainerSession`](https://github.com/OWASP/www-project-csrfguard/blob/master/csrfguard-extensions/csrfguard-extension-session/src/main/java/org/owasp/csrfguard/session/ContainerSession.java).

If your web application is stateful and uses container sessions (_JSESSIONID_), then you don't need to change anything, except adding the [`csrfguard-extension-session`](https://github.com/OWASP/www-project-csrfguard/tree/master/csrfguard-extensions/csrfguard-extension-session) dependency (see https://github.com/OWASP/www-project-csrfguard#using-with-maven), where the reference implementation is stored.

Please see the [bundled test application](https://github.com/OWASP/www-project-csrfguard/tree/master/csrfguard-test/csrfguard-test-jsp) for an example.

If your webapp is stateless (e.g. using _JWTs_), then you'll have to implement your own `LogicalSession` and `LogicalSessionExtractor`, where you need to specify how to identify and extract the actor from a request, how to terminate the "user session" and so on.
